utils/kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/validator/TopicsExistenceValidator.scala

```diff
@@ -1,99 +1,35 @@
-    validateTopics(List(topic)).map(_.head)
+  final def validateTopic[T <: TopicName](topic: T): Validated[TopicExistenceValidationException[T], T] =
+    validateTopics(NonEmptyList.one(topic)).map(_.head)
+
+  def validateTopics[T <: TopicName](
+      topics: NonEmptyList[T]
+  ): Validated[TopicExistenceValidationException[T], NonEmptyList[T]]
 
-  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]]
 }
 
-trait WithCachedTopicsExistenceValidator extends TopicsExistenceValidator {
-  protected val kafkaConfig: KafkaConfig
-  protected lazy val validator = new CachedTopicsExistenceValidator(kafkaConfig)
+final case class TopicExistenceValidationException[T <: TopicName](topics: NonEmptyList[T])
+    extends RuntimeException(TopicExistenceValidationException.message(topics)) {
 
-  final override def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] =
-    validator.validateTopics(topics)
+  def toCustomNodeError(nodeId: String, paramName: Option[ParameterName]) =
+    new CustomNodeError(nodeId, super.getMessage, paramName)
 }
 
-class CachedTopicsExistenceValidator(kafkaConfig: KafkaConfig) extends TopicsExistenceValidator with LazyLogging {
-  private def config = kafkaConfig.topicsExistenceValidationConfig.validatorConfig
-  @transient private lazy val autoCreateSettingCache =
-    new SingleValueCache[Boolean](expireAfterAccess = None, expireAfterWrite = Some(config.autoCreateFlagFetchCacheTtl))
-  @transient private lazy val topicListCache =
-    new SingleValueCache[List[String]](expireAfterAccess = None, expireAfterWrite = Some(config.topicsFetchCacheTtl))
+object TopicExistenceValidationException {
 
-  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] = {
-    if (!kafkaConfig.topicsExistenceValidationConfig.enabled || isAutoCreateEnabled()) {
-      Valid(topics)
-    } else {
-      topicListCache
-        .get()
-        .flatMap(existingTopics => {
-          if (topics.diff(existingTopics).isEmpty)
-            Some(Valid(topics))
-          else
-            None
-        })
-        .getOrElse {
-          val existingTopics = usingAdminClient {
-            _.listTopics(new ListTopicsOptions().timeoutMs(config.adminClientTimeout.toMillis.toInt))
-              .names()
-              .get()
-              .asScala
-              .toList
-          }
-          topicListCache.put(existingTopics)
-          val notExistingTopics = topics.diff(existingTopics)
-          if (notExistingTopics.isEmpty)
-            Valid(topics)
-          else
-            Invalid(new TopicExistenceValidationException(notExistingTopics))
-        }
+  private[TopicExistenceValidationException] def message[T <: TopicName](topics: NonEmptyList[T]): String =
+    topics.tail match {
+      case Nil => s"Topic ${topics.head.toUnspecialized.name} doesn't not exist"
```

Comment:
I intentionally didn't add the name val/def to the trait. I want devs to pattern match on this type instead of using TopicName without specialized context. In this case probably `UnspecializedTopicName` should be used instead. Here I want to do it in one line and pattern matching is not so convenient to use in one-liners, so I converted to `UnspecializedTopicName` before. 

I also think that changing data class definitions only for the sake of easiness of rendering strings inside logs or exceptions should not be considered.

Reply:
Good points, hadn't though of that, thanks :+1: